x-pack/plugins/enterprise_search/server/lib/enterprise_search_request_handler.ts

```diff
@@ -7,63 +7,90 @@
-    response: KibanaResponseFactory
-  ) => {
-    try {
-      const enterpriseSearchUrl = config.host as string;
-      const params = request.query ? `?${querystring.stringify(request.query)}` : '';
-      const url = `${encodeURI(enterpriseSearchUrl)}${path}${params}`;
+export class EnterpriseSearchRequestHandler {
+  private enterpriseSearchUrl: string;
+  private log: Logger;
 
-      const apiResponse = await fetch(url, {
-        headers: { Authorization: request.headers.authorization as string },
-      });
+  constructor({ config, log }: IConstructorDependencies) {
+    this.log = log;
+    this.enterpriseSearchUrl = config.host as string;
+  }
 
-      const body = await apiResponse.json();
+  createRequest<ResponseBody>({
+    path,
+    params,
+    hasValidData = () => true,
+  }: IRequestParams<ResponseBody>) {
+    return async (
+      _context: RequestHandlerContext,
+      request: KibanaRequest<unknown, Readonly<{}>, unknown>,
+      response: KibanaResponseFactory
+    ) => {
+      try {
+        // Set up API URL
+        params = params ?? (request.query ? `?${querystring.stringify(request.query)}` : '');
+        const url = encodeURI(this.enterpriseSearchUrl + path + params);
 
-      if (hasValidData(body)) {
-        return response.ok({ body });
-      } else {
-        throw new Error(`Invalid data received: ${JSON.stringify(body)}`);
-      }
-    } catch (e) {
-      log.error(`Cannot connect to Enterprise Search: ${e.toString()}`);
-      if (e instanceof Error) log.debug(e.stack as string);
+        // Set up API options
+        const { method } = request.route;
+        const headers = { Authorization: request.headers.authorization as string };
+        const body = Object.keys(request.body as object).length
+          ? JSON.stringify(request.body)
+          : undefined;
+
+        // Call the Enterprise Search API and pass back response to the front-end
+        const apiResponse = await fetch(url, { method, headers, body });
+
+        if (apiResponse.url.endsWith('/login')) {
+          throw new Error('Cannot authenticate Enterprise Search user');
+        }
 
-      return response.customError({
-        statusCode: 502,
-        body: 'Error connecting or fetching data from Enterprise Search',
-      });
-    }
-  };
+        const { status } = apiResponse;
+        const json = await apiResponse.json();
+
+        if (hasValidData(json)) {
+          return response.custom({ statusCode: status, body: json });
+        } else {
+          this.log.debug(`Invalid data received from <${url}>: ${JSON.stringify(json)}`);
+          throw new Error('Invalid data received');
+        }
+      } catch (e) {
+        const errorMessage = `Error connecting to Enterprise Search: ${e?.message || e.toString()}`;
+
+        this.log.error(errorMessage);
+        if (e instanceof Error) this.log.debug(e.stack as string);
+
+        return response.customError({ statusCode: 502, body: errorMessage });
```

Comment:
😁 I was accidentally able to test the new functionality where we send custom error messages back to the console (instead of the previous generic "Error connecting" message) when implementing the `body` payload:

<img width="727" alt="Screen Shot 2020-08-26 at 4 53 40 PM" src="https://user-images.githubusercontent.com/549407/91466072-79139200-e843-11ea-8708-c8a9e6b96ef8.png">

VERY useful for debugging, and I'm happy w/ this change - the only thing to keep an eye on is if somehow start catching errors that leak sensitive info in the message somehow.